src/mixins/CoWSwapOnchainOrders.sol

```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8;
+
+import "../vendored/GPv2Order.sol";
+import "../interfaces/ICoWSwapOnchainOrders.sol";
+import "../libraries/CoWSwapEip712.sol";
+
+/// @title CoW Swap Onchain Order Creator Event Emitter
+/// @author CoW Swap Developers
+contract CoWSwapOnchainOrders is ICoWSwapOnchainOrders {
+    using GPv2Order for GPv2Order.Data;
+    using GPv2Order for bytes;
+
+    /// @dev The domain separator for the CoW Swap settlement contract.
+    bytes32 internal immutable cowSwapDomainSeparator;
+
+    /// @param settlementContractAddress The address of CoW Swap's settlement contract on the chain where this contract
+    /// is deployed.
+    constructor(address settlementContractAddress) {
+        cowSwapDomainSeparator = CoWSwapEip712.domainSeparator(
+            settlementContractAddress
+        );
+    }
+
+    /// @dev Emits an event with all information needed to execute an order onchain and returns the corresponding order
+    /// hash.
+    ///
+    /// See [`ICoWSwapOnchainOrders.OrderPlacement`] for details on the meaning of each parameter.
+    /// @return The EIP-712 hash of the order data as computed by the CoW Swap settlement contract.
+    function broadcastOrder(
+        address sender,
+        GPv2Order.Data memory order,
+        OnchainSignature memory signature,
+        bytes memory data
+    ) internal returns (bytes32) {
+        emit OrderPlacement(sender, order, signature, data);
+        return order.hash(cowSwapDomainSeparator);
```

Comment:
Maybe this is a silly question, but why return the order hash in favour of including it with the event? For the most part I consider returning this hash only practically useful if it is used within the transaction, but what I gather from the design, it's more practically useful for the event indexer. From a dune standpoint this would require a join of event data with call and imagine the same for populating our orderbook. 

Am I missing something here?

Reply:
In principle, it allows you to do stuff like:
```
bytes memory orderUid = onchainOrders.broadcastOrder(...);
settlement.setPreSignature(orderUid, true);
```

Reply:
In practice, I expect most contracts using this event emitter to also need the order hash, like in the case of the eth flow, since it's a unique way to represent this order (excluding the owner). The original idea was to include the full orderUid, but this is redundant data that require extra memory allocation, so overall I thought it wasn't worth it since it's easy to compute.

There is a point to make that if this function is called outside of a contract (e.g., an EOA that wants to create an event onchain) this is information that isn't needed. On the other hand, I think the current implementation is good for being inherited by other contracts.